In [1]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
from datetime import datetime, timedelta
from sklearn.model_selection import train_test_split


# 1. Data preprocessing

In [28]:
data_df = pd.read_csv('CreditTraining.csv')
label_df = data_df['Prod_Category'].tolist()

In [29]:
# data_df = data_df.drop(['Prod_Category'], axis=1)
data_df.head()

Id_Customer  Y        Customer_Type   BirthDate Customer_Open_Date  \
0         7440  0  Non Existing Client  07/08/1977         13/02/2012   
1          573  0      Existing Client  13/06/1974         04/02/2009   
2         9194  0  Non Existing Client  07/11/1973         03/04/2012   
3         3016  1      Existing Client  08/07/1982         25/08/2011   
4         6524  0  Non Existing Client  18/08/1953         10/01/2012   

    P_Client Educational_Level Marital_Status  Number_Of_Dependant  \
0  NP_Client        University        Married                  3.0   
1   P_Client        University        Married                  0.0   
2  NP_Client        University        Married                  2.0   
3  NP_Client        University        Married                  3.0   
4  NP_Client        University        Married                  2.0   

   Years_At_Residence Net_Annual_Income  Years_At_Business Prod_Sub_Category  \
0                   1                36                1.0                 C   
1                  12                18                2.0                 C   
2                  10                36                1.0                 C   
3                   3                36                1.0                 C   
4                   1                36                1.0                 C   

  Prod_Decision_Date Source Type_Of_Residence  Nb_Of_Products  \
0         14/02/2012  Sales             Owned               1   
1         30/06/2011  Sales           Parents               1   
2         04/04/2012  Sales             Owned               1   
3         07/09/2011  Sales          New rent               1   
4         11/01/2012  Sales             Owned               1   

  Prod_Closed_Date Prod_Category  
0              NaN             B  
1              NaN             G  
2              NaN             B  
3       31/12/2012             L  
4              NaN             D

Observations: there are categorical data, and we would to deal with that 

## 1.0. Types of data

In [30]:
clean_df = data_df.select_dtypes(exclude=['object']).copy()

In [31]:
cat_df = data_df.select_dtypes(include=['object']).copy()
cat_df.head()

Customer_Type   BirthDate Customer_Open_Date   P_Client  \
0  Non Existing Client  07/08/1977         13/02/2012  NP_Client   
1      Existing Client  13/06/1974         04/02/2009   P_Client   
2  Non Existing Client  07/11/1973         03/04/2012  NP_Client   
3      Existing Client  08/07/1982         25/08/2011  NP_Client   
4  Non Existing Client  18/08/1953         10/01/2012  NP_Client   

  Educational_Level Marital_Status Net_Annual_Income Prod_Sub_Category  \
0        University        Married                36                 C   
1        University        Married                18                 C   
2        University        Married                36                 C   
3        University        Married                36                 C   
4        University        Married                36                 C   

  Prod_Decision_Date Source Type_Of_Residence Prod_Closed_Date Prod_Category  
0         14/02/2012  Sales             Owned              NaN             B  
1         30/06/2011  Sales           Parents              NaN             G  
2         04/04/2012  Sales             Owned              NaN             B  
3         07/09/2011  Sales          New rent       31/12/2012             L  
4         11/01/2012  Sales             Owned              NaN             D

In [32]:
cat_df.columns

Index(['Customer_Type', 'BirthDate', 'Customer_Open_Date', 'P_Client',
       'Educational_Level', 'Marital_Status', 'Net_Annual_Income',
       'Prod_Sub_Category', 'Prod_Decision_Date', 'Source',
       'Type_Of_Residence', 'Prod_Closed_Date', 'Prod_Category'],
      dtype='object')

In [33]:
cat_df['Marital_Status'].value_counts()

Married      4206
Single       1046
Widowed        64
Divorced       63
Separated       1
Name: Marital_Status, dtype: int64

## 1.1. Deal with categorical data excluding date

From all categorical datas ['Customer_Type', 'BirthDate', 'Customer_Open_Date', 'P_Client',
       'Educational_Level', 'Marital_Status', 'Net_Annual_Income',
       'Prod_Sub_Category', 'Prod_Decision_Date', 'Source',
       'Type_Of_Residence', 'Prod_Closed_Date', 'Prod_Category']

We first exclude ['BirthDate', 'Customer_Open_Date', 'Net_Annual_Income', 'Prod_Decision_Date']

We use 0-1 encoding for each category

In [34]:
for name in ['Customer_Type', 'P_Client',
            'Educational_Level', 'Marital_Status',
            'Prod_Sub_Category', 'Source',
            'Type_Of_Residence', 'Prod_Category']:
        print(cat_df[name].value_counts())

Non Existing Client    3369
Existing Client        2011
Name: Customer_Type, dtype: int64
NP_Client    4968
P_Client      412
Name: P_Client, dtype: int64
University           4785
Master/PhD            522
Diploma                58
Secondary or Less      15
Name: Educational_Level, dtype: int64
Married      4206
Single       1046
Widowed        64
Divorced       63
Separated       1
Name: Marital_Status, dtype: int64
C    4638
G     624
P     118
Name: Prod_Sub_Category, dtype: int64
Sales     4119
Branch    1261
Name: Source, dtype: int64
Owned       4791
Old rent     323
Parents      179
New rent      83
Company        4
Name: Type_Of_Residence, dtype: int64
B    3176
D     670
C     517
K     265
L     236
G     188
E     101
H      79
J      71
M      49
A      19
F       5
I       4
Name: Prod_Category, dtype: int64


In [35]:
pd.get_dummies(cat_df[['Customer_Type', 'P_Client',
                       'Educational_Level', 'Marital_Status',
                       'Prod_Sub_Category', 'Source',
                       'Type_Of_Residence', 'Prod_Category']])[:10]

Customer_Type_Existing Client  Customer_Type_Non Existing Client  \
0                              0                                  1   
1                              1                                  0   
2                              0                                  1   
3                              1                                  0   
4                              0                                  1   
5                              1                                  0   
6                              0                                  1   
7                              0                                  1   
8                              1                                  0   
9                              1                                  0   

   P_Client_NP_Client  P_Client_P_Client  Educational_Level_Diploma  \
0                   1                  0                          0   
1                   0                  1                          0   
2                   1                  0                          0   
3                   1                  0                          0   
4                   1                  0                          0   
5                   1                  0                          0   
6                   1                  0                          0   
7                   1                  0                          0   
8                   1                  0                          0   
9                   1                  0                          0   

   Educational_Level_Master/PhD  Educational_Level_Secondary or Less  \
0                             0                                    0   
1                             0                                    0   
2                             0                                    0   
3                             0                                    0   
4                             0                                    0   
5                             0                                    0   
6                             0                                    0   
7                             0                                    0   
8                             0                                    0   
9                             0                                    0   

   Educational_Level_University  Marital_Status_Divorced  \
0                             1                        0   
1                             1                        0   
2                             1                        0   
3                             1                        0   
4                             1                        0   
5                             1                        0   
6                             1                        0   
7                             1                        0   
8                             1                        0   
9                             1                        0   

   Marital_Status_Married  ...  Prod_Category_D  Prod_Category_E  \
0                       1  ...                0                0   
1                       1  ...                0                0   
2                       1  ...                0                0   
3                       1  ...                0                0   
4                       1  ...                1                0   
5                       1  ...                0                0   
6                       1  ...                0                0   
7                       1  ...                0                0   
8                       0  ...                0                1   
9                       1  ...                0                0   

   Prod_Category_F  Prod_Category_G  Prod_Category_H  Prod_Category_I  \
0                0                0                0                0   
1                0                1                0                0   
2                

In [36]:
clean_df1 = clean_df.join(pd.get_dummies(cat_df[['Customer_Type', 'P_Client',
                                                'Educational_Level', 'Marital_Status',
                                                'Prod_Sub_Category', 'Source',
                                                'Type_Of_Residence', 'Prod_Category']]))

## 1.2. Deal with years

['BirthDate', 'Customer_Open_Date', 'Net_Annual_Income', 'Prod_Decision_Date', 'Prod_Closed_Date']

- **Possible appraoches** 
    - Use duration
        - `Birth_Duration = Now - BirthDate` ('BirthDate'): assuming there is a distribution of credibility, just started working -> less credit, worked for a long time but not close to retirement -> high credit) 
        - `Customer_Open_Duration = Now - Customer_Open_Date` ('Customer_Open_Date'): usually longer the history is, the more royal the customer is
        - 'Prod_Closed_Date' - 'Prod_Decision_Date'
            - length of the product?: if product closed 
            - `Prod_not_closed = (Prod_closed != nan)`: dummy variable if the product is closed 'Prod_not_closed' = 0 if closed
            - `Prod_Decision_Duration = Now - Prod_Decision_Date`
            - `Prod_Closed_Duration = Now - Prod_Closed_Date`: inf if not closed
    - Set severals intervals 
        - 'Net_Annual_Income': 


In [37]:
type(cat_df['Prod_Closed_Date'].tolist()[3]) == float


False

In [38]:
# del clean_df1['BirthDate']
# del clean_df1['Customer_Open_Date'] 
# del clean_df1['Prod_Decision_Date']
# del clean_df1['Prod_closed']


In [39]:
# clean_df1['Prod_Decision_Duration'].tolist()[1].days

In [40]:
# duration: birth - now
temp_list = [(datetime.now().date() - datetime.strptime(datetime_str, '%d/%m/%Y').date()).days for datetime_str in cat_df['BirthDate'].tolist()]
clean_df1['Birth_Duration'] = temp_list

# duration: Customer_Open_Date
temp_list = [(datetime.now().date() - datetime.strptime(datetime_str, '%d/%m/%Y').date()).days for datetime_str in cat_df['Customer_Open_Date'].tolist()]
clean_df1['Customer_Open_Duration'] = temp_list

# dummy var: product closed = 1
temp_list = [int(type(ele) != float) for ele in cat_df['Prod_Closed_Date'].tolist()]
clean_df1['Prod_not_closed'] = temp_list

# duration: Prod_Decision_Date
temp_list = [(datetime.now().date() - datetime.strptime(datetime_str, '%d/%m/%Y').date()).days for datetime_str in cat_df['Prod_Decision_Date'].tolist()]
clean_df1['Prod_Decision_Duration'] = temp_list



# # duration: Prod_Closed_Date
# temp_list = [int(type(ele) != float) for ele in cat_df['Prod_Closed_Date'].tolist()]
# clean_df1['Prod_closed'] = temp_list






In [41]:
clean_df1

1                              0   
5363                1.0               1                              1   
5364                1.0               1                              1   
5365                2.0               1                              0   
5366                1.0               2                              0   
5367               10.0               1                              0   
5368                3.0               1                              0   
5369                1.0               1                              1   
5370                1.0               1                              1   
5371                7.0               1                              1   
5372                3.0               1                              0   
5373                1.0               1                              1   
5374                1.0               1                              0   
5375                1.0               1                              0   
5376                1.0               1                              1   
5377               30.0               1                              0   
5378                1.0               1                              0   
5379                1.0               1                              0   

      Customer_Type_Non Existing Client  P_Client_NP_Client  \
0                                     1                   1   
1                                     0                   0   
2                                     1                   1   
3                                     0                   1   
4                                     1                   1   
5                                     0                   1   
6                                     1                   1   
7                                     1                   1   
8                                     0                   1   
9                                     0                   1   
10                                    1                   1   
11                                    1                   1   
12                                    0                   0   
13                                    1                   1   
14                                    1                   1   
15                                    1                   1   
16                                    1                   1   
17                                    0                   1   
18                                    1                   1   
19                                    1                   1   
20                                    1                   1   
21                                    1                   1   
22                                    1                   1   
23                                    0                   0   
24                                    1                   1   
25                                    1                   1   
26                                    1                   1   
27                                    1                   1   
28                                    0                   1   
29                                    1                   1   
...                                 ...                 ...   
5350                                  1                   1   
5351                                  1                   1   
5352                                  0                   0   
5353                                  0                   1   
5354                                  0                   1   
5355                                  0                   1   
5356                                  1                   1   
5357                                  0                   0   
5358                                  1                   1   
5359                                  0                   1   
5360                                  1                   1

## 1.3. Split train and test sets 

In [59]:
clean_df1.columns.values

array(['Id_Customer', 'Y', 'Number_Of_Dependant', 'Years_At_Residence',
       'Years_At_Business', 'Nb_Of_Products',
       'Customer_Type_Existing Client',
       'Customer_Type_Non Existing Client', 'P_Client_NP_Client',
       'P_Client_P_Client', 'Educational_Level_Diploma',
       'Educational_Level_Master/PhD',
       'Educational_Level_Secondary or Less',
       'Educational_Level_University', 'Marital_Status_Divorced',
       'Marital_Status_Married', 'Marital_Status_Separated',
       'Marital_Status_Single', 'Marital_Status_Widowed',
       'Prod_Sub_Category_C', 'Prod_Sub_Category_G',
       'Prod_Sub_Category_P', 'Source_Branch', 'Source_Sales',
       'Type_Of_Residence_Company', 'Type_Of_Residence_New rent',
       'Type_Of_Residence_Old rent', 'Type_Of_Residence_Owned',
       'Type_Of_Residence_Parents', 'Prod_Category_A', 'Prod_Category_B',
       'Prod_Category_C', 'Prod_Category_D', 'Prod_Category_E',
       'Prod_Category_F', 'Prod_Category_G', 'Prod_Category_H',
 

In [42]:
data_list = []
labels = []
for name in clean_df1.columns.values:
    if name is 'Id_Customer':
        pass
    elif name is 'Y':
        labels = clean_df1[name].tolist()
    else:
        data_list.append(clean_df1[name].tolist())

data_list = np.transpose(data_list)



## 1.4. Check data

## 1.4.1. Nan elemnts

In [43]:
np.shape(data_list), np.shape(clean_df1)

((5380, 45), (5380, 46))

In [52]:
# check nan elements
np.where(np.isnan(X_train))

(array([  23,  887, 2126, 2405]), array([1, 3, 3, 1]))

In [66]:
# nan in firt and third col 
sum(np.isnan(clean_df1['Number_Of_Dependant'].tolist())), sum(np.isnan(clean_df1['Years_At_Business'].tolist()))


(2, 2)

In [70]:
clean_df1[np.isnan(clean_df1['Number_Of_Dependant'].tolist())]

Id_Customer  Y  Number_Of_Dependant  Years_At_Residence  \
634          8953  1                  NaN                  21   
5045         9588  0                  NaN                  10   

      Years_At_Business  Nb_Of_Products  Customer_Type_Existing Client  \
634                 1.0               2                              1   
5045               10.0               1                              1   

      Customer_Type_Non Existing Client  P_Client_NP_Client  \
634                                   0                   0   
5045                                  0                   1   

      P_Client_P_Client  ...  Prod_Category_H  Prod_Category_I  \
634                   1  ...                0                0   
5045                  0  ...                0                0   

      Prod_Category_J  Prod_Category_K  Prod_Category_L  Prod_Category_M  \
634                 0                0                0                0   
5045                0                0                0                0   

      Birth_Duration  Customer_Open_Duration  Prod_not_closed  \
634            10899                    3489                1   
5045           25481                    5682                0   

      Prod_Decision_Duration  
634                     2913  
5045                    2905  

[2 rows x 46 columns]

In [71]:
clean_df1[np.isnan(clean_df1['Years_At_Business'].tolist())]

Id_Customer  Y  Number_Of_Dependant  Years_At_Residence  \
1987          398  0                  0.0                  10   
2750         5882  0                  0.0                   2   

      Years_At_Business  Nb_Of_Products  Customer_Type_Existing Client  \
1987                NaN               2                              0   
2750                NaN               1                              0   

      Customer_Type_Non Existing Client  P_Client_NP_Client  \
1987                                  1                   1   
2750                                  1                   1   

      P_Client_P_Client  ...  Prod_Category_H  Prod_Category_I  \
1987                  0  ...                0                0   
2750                  0  ...                0                0   

      Prod_Category_J  Prod_Category_K  Prod_Category_L  Prod_Category_M  \
1987                0                0                0                0   
2750                0                0                0                0   

      Birth_Duration  Customer_Open_Duration  Prod_not_closed  \
1987           13114                    3248                1   
2750           14124                    3039                0   

      Prod_Decision_Duration  
1987                    3241  
2750                    3026  

[2 rows x 46 columns]

##### Customer with nan 
- Years_At_Business
    - Id_Customer = 398, 5882	
- Number_Of_Dependant
    - Id_Customer = 8953, 9588	



##### Potential approach:
- Delete the data 
- Replace value with the median 

In [72]:
clean_df2 = clean_df1.copy

temp_list[] = clean_df2['Number_Of_Dependant']
temp_list[] = 
clean_df2['Number_Of_Dependant'] = 



In [45]:
X_train, X_test, y_train, y_test = train_test_split(data_list, labels, test_size=0.25, random_state=42)

np.save('X_train.npy',X_train)
np.save('X_test.npy',X_test)
np.save('y_train.npy',y_train)
np.save('y_test.npy',y_test)

# 2. Deterministic Models

This is a binary classification problem where we only have two labels 

In [22]:
for i in set(labels):
    print(i,labels.count(i))

0 4987
1 393


## 2.1. KNN


In [47]:
from sklearn.neighbors import KNeighborsClassifier

neigh = KNeighborsClassifier(n_neighbors=2)
neigh.fit(X_train, y_train)
pred_KNN = neigh.predict(X_test)

ValueError: Input contains NaN, infinity or a value too large for dtype('float64').